File: brjobs-java/src/main/java/ads/uninassau/brjobs/model/Usuario.java
package ads.uninassau.brjobs.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Entidade que representa um usuário no sistema.
 * Pode ser um CONTRATANTE ou um PRESTADOR de serviços.
 * Se for PRESTADOR, terá um relacionamento com a entidade Prestador.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "usuarios")
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String nome;

    @Column(unique = true, nullable = false)
    private String email;

    @Column(nullable = false)
    private String senha;

    @Column
    private String telefone;

    @Column(unique = true)
    private String cpf;

    @Column
    private String genero;

    @Column(name = "data_nascimento")
    private LocalDate dataNascimento;

    @Column
    private String endereco;

    @Column(name = "cep", length = 20)
    private String cep;

    @Column(name = "rua")
    private String rua;

    @Column(name = "bairro")
    private String bairro;

    @Column(name = "cidade")
    private String cidade;

    @Column(name = "estado", length = 2)
    private String estado;

    @Column(name = "numero")
    private String numero;

    @Column(name = "complemento")
    private String complemento;

    @Column(name = "bio", length = 600)
    private String bio;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private TipoUsuario tipoUsuario;

    @Column(nullable = false)
    private boolean ativo = true;

    @CreationTimestamp
    @Column(name = "data_cadastro", nullable = false, updatable = false)
    private LocalDateTime dataCadastro;

    @Column(name = "data_atualizacao")
    private LocalDateTime dataAtualizacao;

    @Column(name = "password_reset_code", length = 6)
    private String passwordResetCode;

    // Hash (SHA-256) do token de redefinição enviado no link do e-mail.
    // Guardamos só o hash: vazamento do banco não expõe tokens utilizáveis.
    @Column(name = "password_reset_token_hash", length = 64)
    private String passwordResetTokenHash;

    @Column(name = "password_reset_expires_at")
    private LocalDateTime passwordResetExpiresAt;

    // Confirmação de e-mail no cadastro (não bloqueia login; registra a confirmação).
    @Column(name = "email_confirmado")
    private Boolean emailConfirmado;

    @Column(name = "email_confirmation_token", length = 64)
    private String emailConfirmationToken;

    @Column(name = "email_confirmation_expires_at")
    private LocalDateTime emailConfirmationExpiresAt;

    @PreUpdate
    protected void onUpdate() {
        dataAtualizacao = LocalDateTime.now();
    }

    // Foto de perfil armazenada como bytea
    @Column(name = "foto_perfil")
    private byte[] fotoPerfil;

    // Relacionamento com Prestador (opcional, apenas se tipoUsuario == PRESTADOR)
    @OneToOne(mappedBy = "usuario", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @ToString.Exclude
    private Prestador prestador;

    /**
     * Verifica se o usuário é um prestador de serviço
     */
    public boolean isPrestador() {
        return tipoUsuario.isPrestador();
    }

    /**
     * Verifica se o usuário é um contratante
     */
    public boolean isContratante() {
        return tipoUsuario.isContratante();
    }

    /**
     * Verifica se a conta foi criada via login social (Google/Facebook/Apple)
     * e ainda não definiu uma senha local. Contas sociais recebem um
     * placeholder na coluna senha (NOT NULL), nunca um hash BCrypt.
     */
    public boolean isSomenteLoginSocial() {
        return senha == null
                || "SOCIAL_LOGIN".equals(senha)
                || senha.startsWith("OAUTH2_");
    }

    /**
     * Ativa o usuário
     */
    public void ativar() {
        this.ativo = true;
    }

    /**
     * Desativa o usuário
     */
    public void desativar() {
        this.ativo = false;
    }
}

File: brjobs-flutter/lib/features/auth/domain/usuario.dart
import '../../../shared/utils/json_utils.dart';
import 'tipo_usuario.dart';

/// Espelha `UsuarioDTO` do backend.
class Usuario {
  const Usuario({
    required this.id,
    this.tipoUsuario,
    required this.nome,
    required this.email,
    this.telefone,
    this.endereco,
    this.cep,
    this.rua,
    this.bairro,
    this.cidade,
    this.estado,
    this.numero,
    this.complemento,
    this.bio,
    this.cpf,
    this.genero,
    this.dataNascimento,
    this.dataCadastro,
    this.ativo = true,
    this.fotoPerfilUrl,
  });

  final int id;
  final TipoUsuario? tipoUsuario;
  final String nome;
  final String email;
  final String? telefone;
  final String? endereco;
  final String? cep;
  final String? rua;
  final String? bairro;
  final String? cidade;
  final String? estado;
  final String? numero;
  final String? complemento;
  final String? bio;
  final String? cpf;
  final String? genero;
  final DateTime? dataNascimento;
  final DateTime? dataCadastro;
  final bool ativo;

  /// Não faz parte do `UsuarioDTO`, mas o login social devolve `fotoPerfil`.
  final String? fotoPerfilUrl;

  bool get isPrestador => tipoUsuario?.isPrestador ?? false;

  factory Usuario.fromJson(Map<String, dynamic> json) {
    return Usuario(
      id: asInt(json['id']) ?? 0,
      tipoUsuario: TipoUsuario.fromWire(json['tipoUsuario']?.toString()),
      nome: json['nome']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
      telefone: json['telefone']?.toString(),
      endereco: json['endereco']?.toString(),
      cep: json['cep']?.toString(),
      rua: json['rua']?.toString(),
      bairro: json['bairro']?.toString(),
      cidade: json['cidade']?.toString(),
      estado: json['estado']?.toString(),
      numero: json['numero']?.toString(),
      complemento: json['complemento']?.toString(),
      bio: json['bio']?.toString(),
      cpf: json['cpf']?.toString(),
      genero: json['genero']?.toString(),
      dataNascimento: asDate(json['dataNascimento']),
      dataCadastro: asDate(json['dataCadastro']),
      ativo: asBool(json['ativo'], fallback: true),
      fotoPerfilUrl:
          json['fotoPerfil']?.toString() ?? json['fotoPerfilUrl']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => pruneNulls({
        'id': id,
        'tipoUsuario': tipoUsuario?.wire,
        'nome': nome,
        'email': email,
        'telefone': telefone,
        'endereco': endereco,
        'cep': cep,
        'rua': rua,
        'bairro': bairro,
        'cidade': cidade,
        'estado': estado,
        'numero': numero,
        'complemento': complemento,
        'bio': bio,
        'cpf': cpf,
        'genero': genero,
        'dataNascimento': toLocalDate(dataNascimento),
        'ativo': ativo,
      });

  Usuario copyWith({
    String? nome,
    String? telefone,
    String? endereco,
    String? cep,
    String? rua,
    String? bairro,
    String? cidade,
    String? estado,
    String? numero,
    String? complemento,
    String? bio,
    String? genero,
    DateTime? dataNascimento,
    String? fotoPerfilUrl,
  }) {
    return Usuario(
      id: id,
      tipoUsuario: tipoUsuario,
      nome: nome ?? this.nome,
      email: email,
      telefone: telefone ?? this.telefone,
      endereco: endereco ?? this.endereco,
      cep: cep ?? this.cep,
      rua: rua ?? this.rua,
      bairro: bairro ?? this.bairro,
      cidade: cidade ?? this.cidade,
      estado: estado ?? this.estado,
      numero: numero ?? this.numero,
      complemento: complemento ?? this.complemento,
      bio: bio ?? this.bio,
      cpf: cpf,
      genero: genero ?? this.genero,
      dataNascimento: dataNascimento ?? this.dataNascimento,
      dataCadastro: dataCadastro,
      ativo: ativo,
      fotoPerfilUrl: fotoPerfilUrl ?? this.fotoPerfilUrl,
    );
  }
}
